libp2p/network/swarm.py

```diff
@@ -636,17 +683,59 @@ async def close_peer(self, peer_id: ID) -> None:
             except Exception as e:
                 logger.warning(f"Error closing connection to {peer_id}: {e}")
 
+        # Release stream resources for this peer
+        if self._resource_manager is not None:
+            # Release all streams for this peer (both inbound and outbound)
+            # Note: This is a simplified approach - in a real implementation,
+            # we would track individual streams and release them specifically
+            logger.debug("Releasing stream resources for peer %s", peer_id)
+
         # Remove from connections dict
         self.connections.pop(peer_id, None)
 
         logger.debug("successfully close the connection to peer %s", peer_id)
 
-    async def add_conn(self, muxed_conn: IMuxedConn) -> SwarmConn:
+    async def add_conn(self, muxed_conn: IMuxedConn) -> "SwarmConn":
         """
         Add a `IMuxedConn` to `Swarm` as a `SwarmConn`, notify "connected",
         and start to monitor the connection for its new streams and
         disconnection.
         """
+        # If QUIC and resource manager is present, attach a connection scope
+        if (
+            isinstance(muxed_conn, QUICConnection)
+            and self._resource_manager is not None
+        ):
+            try:
+                direction = (
+                    Direction.OUTBOUND
+                    if muxed_conn._is_initiator
+                    else Direction.INBOUND
+                )
+                endpoint = getattr(muxed_conn, "_maddr", None)
+                peer_id_for_scope = muxed_conn.peer_id
+                direction_str = (
+                    "inbound" if direction == Direction.INBOUND else "outbound"
+                )
+                conn_scope = self._resource_manager.open_connection(
```

Comment:
`open_connection` returns a `bool` here, which does not seem to carry enough context to be referred to as a "scope". The only place I can find any definition of what a "resource scope" should look like is within the `tests` folder.